src/ray/object_manager/plasma/create_request_queue.cc

```diff
@@ -83,38 +83,33 @@ std::pair<PlasmaObject, PlasmaError> CreateRequestQueue::TryRequestImmediately(
 bool CreateRequestQueue::ProcessRequest(std::unique_ptr<CreateRequest> &request) {
   // Return an OOM error to the client if we have hit the maximum number of
   // retries.
-  // TODO(sang): Delete this logic?
-  bool evict_if_full = evict_if_full_;
-  if (max_retries_ == 0) {
-    // If we cannot retry, then always evict on the first attempt.
-    evict_if_full = true;
-  } else if (num_retries_ > 0) {
-    // Always try to evict after the first attempt.
-    evict_if_full = true;
-  }
-  request->error = request->create_callback(evict_if_full, &request->result);
+  request->error =
+      request->create_callback(/*evict_if_full=*/evict_if_full_, &request->result);
   return request->error != PlasmaError::OutOfMemory;
 }
 
 Status CreateRequestQueue::ProcessRequests() {
   while (!queue_.empty()) {
     auto request_it = queue_.begin();
     auto create_ok = ProcessRequest(*request_it);
+    auto now = absl::GetCurrentTimeNanos();
     if (create_ok) {
       FinishRequest(request_it);
+      last_success_ns_ = now;
```

Comment:
Why do we need to assign success here?

Reply:
Otherwise it can OOM without any grace period if the spilling is not invoked. (since the last success_ns is not renewed at all)

Then 
```
We need a grace period since (1) global GC takes a bit of time to
        // kick in
```
is not reflected. Let me know if I am missing something!

Reply:
Shouldn't we be setting last_success_ns_ in the else branch? It seems like we need another state like "waiting for OOM" so we know whether to set vs check the last_success_ns_.

And we should add a unit test for this case because I think it's currently broken:
1. Add a create request that succeeds.
2. More than the oom_grace_period later, add another create request that fails. Check that this doesn't raise OOM <-- I think this will throw OOM right now.

Reply:
We also set it if spilling is successful, so it's set whenever a creation/spill succeeds. If neither has succeeded for the grace period, OOM is raised.

Reply:
I don't think that the spill reset is enough because it won't happen if the spill callback is unsuccessful.

Reply:
If the callback fails, then we are fine to raise OOM immediately-- as no spill or create has succeeded for that time. Once we enter this state the store is full and OOM can be raised without delay.

Reply:
I don't think that's right. "No spill or create has succeeded for that time" could just be because no create request was made during that time. We should start the grace period starting from when global GC is triggered and/or spill fails, not when the last create/spill succeeded. With the semantics that you're suggesting, we could raise OOM immediately just because no create requests were added for a while.

Reply:
So the scenario is the object store is full of objects that cannot be evicted, and a create hasn't happened for a while. I can see that, though it seems a relatively unimportant edge case.

How about we address this in a follow-up unless there can be a simple fix? It might not be worth increasing the code complexity otherwise; we can just document it as a possible failure case.

Reply:
I think the fix is simple enough that we should just do it now. We just start the timer after the first OOM instead of setting it after creation success.

Reply:
Ok, so this means tracking `oom_start_time_ns_` instead of last success right? This value would start off as zero, and we set it to `now` on spill failure if it is currently == 0. It is reset to zero on any success. Does that sound right?

Reply:
Yeah, any success or when OOM is actually thrown to the client.

Reply:
You don't want to reset it on OOM right? Otherwise, it will take a long time to OOM many objects (e.g., 10 seconds * N objects)

Reply:
Ah, right.

Reply:
Also note; I am using -1 as a default value of `oom_start_time_ns_ ` instead of 0 because otherwise, tests will fail (since the initial fake clock time is 0).

Reply:
Also, added a test case that checks if objects are successfully created if there was no new object created for a long time.